### PR TITLE
Remove some unnecessary functions from CUDAHooks

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60547 Make pin_memory/is_pinned use BackendSelect
* #60201 Add an optional Device parameter to pin_memory/is_pinned that does nothing
* **#59655 Remove some unnecessary functions from CUDAHooks**

CUDAHooks is to be used solely when you need to call into CUDA
functionality from a context where you cannot directly link to
CUDA libraries.  Neither of hasPrimaryContext nor
getDevceIndexWithPrimaryContext (sic) needs to be used in such
contexts.  By moving them out of CUDAHooks and calling them
directly a dynamic dispatch can be skipped.

I also fixed the typo in getDev(i)ceIndexWithPrimaryContext

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28972946](https://our.internmc.facebook.com/intern/diff/D28972946)

### CODE REVIEW

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28972946).

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28972946).

File: aten/src/ATen/cuda/CUDAGraphsUtils.cuh [hunk 1]
```diff
@@ -3,6 +3,7 @@
 #include <ATen/CUDAGeneratorImpl.h>
 #include <ATen/cuda/CUDAEvent.h>
 #include <ATen/cuda/detail/UnpackRaw.cuh>
+#include <ATen/cuda/detail/CUDAHooks.h>
```
Comment (swolchok) [inline aten/src/ATen/cuda/CUDAGraphsUtils.cuh:6]: nit: alphabetize? (clang-format would solve this)

**Comment by swolchok** [comment]
no complaints but don't have context to review whether the intent of the diff is a good one. would resign as reviewer but apparently GitHub doesn't have that option?

**Review by ngimel — APPROVED** [review]

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28972946).

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28972946).

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28972946).

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28972946).

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**